lib/response.ex

```diff
@@ -1,5 +1,42 @@
 defmodule HTTPServer.Response do
-  defstruct status_code: 404, status_message: :not_found, resource: nil, headers: %{}, body: ""
-  @type status_code :: 200 | 404
-  @type status_message :: :ok | :not_found
+  defstruct status_code: nil, status_message: "", resource: nil, headers: %{}, body: ""
+
+  @type t :: %__MODULE__{
+          status_code: 200 | 404,
+          status_message: String.t(),
+          resource: String.t(),
+          headers: %{optional(String.t()) => any},
+          body: String.t()
+        }
+
+  def build_response(req, status_code) do
+    %__MODULE__{
+      status_code: status_code,
+      status_message: status_message(status_code),
+      resource: req.resource,
+      headers: req.headers,
+      body: req.body
+    }
+  end
```

Comment:
I would expect a response builder to build a response based off of a Response object.  This takes a `req`, that is a bit code smelly.

Something more like this, which also guarantees that `build_response` only works on `%Response`-s
```suggestion
  def build_response(response = %Response{}, status_code) do
    %__MODULE__{
      status_code: status_code,
      status_message: status_message(status_code),
      resource: response.resource,
      headers: response.headers,
      body: response.body
    }
  end
```

Reply:
In other words, should a `%Response` know about a `%Request` at all?

Reply:
Now that you point it out, I agree that it shouldn't. I mentioned below but I'm changing this now so that the only thing passed to build_response is the status code and a maybe a body. I think I'm realizing too that the headers and resource don't necessarily have to be exact replica's of what was sent in the request, they can convey any messages they want as long as it's clear what they mean to the person making the request.